### ISSUE TITLE
gitlab: Take advantage of GraphQL api.

### ISSUE THREAD

**Issue by ryneeverett** [ralphbean/bugwarrior#905]
Idea [contributed in #624](https://github.com/ralphbean/bugwarrior/issues/624#issuecomment-463677702).

We should test it out and explore:

- Can it improve performance?
- Do all gitlab instances necessarily support this API, or do we need to maintain backwards compatibility with the REST API?
- Are hosted gitlab instances likely to exhibit similar performance differences between the GraphQL and REST API's as gitlab.com?

**Comment by shymega** [comment]
As for:

>  Do all gitlab instances necessarily support this API, or do we need to maintain backwards compatibility with the REST API?

I was thinking we could do a fail gracefully approach - first try to probe the GitLab instance for GraphQL support, then fall back to the REST API? We could add an explicit option in the configuration, which would skip the probe.

For my usage of Bugwarrior, I find gitlab.com is fast, but the hosted instance I use on a day-to-day basis is very slow to get results from. I'd like to improve this, but I wasn't sure if there's any work on GraphQL support?

**Comment by ryneeverett** [comment]
Nobody has ever worked on this to my knowledge.

I would say we probably don't need to worry about maintaining REST support. I believe the graphql API was relatively new when I made that comment. Looking at both the [REST](https://docs.gitlab.com/api/rest/) and [graphql](https://docs.gitlab.com/api/graphql/) api docs, they both seem to have the exact same support:

>     Tier: Free, Premium, Ultimate
>     Offering: GitLab.com, GitLab Self-Managed, GitLab Dedicated

I would also say that we don't necessarily need to switch *entirely* to the graphql api -- it's ok if we just target the slowest queries.

**Comment by shymega** [comment]
Thanks. Do we have any stats or a list of slowest queries? There's a broad set of issues relating to this, and it'd be nice to have a breakdown of the slow queries.

**Comment by ryneeverett** [comment]
No, we do not have any data like that. The question highlights the fact that we don't really have a method to compare performance. We've basically only gone after such low-hanging fruit that the question has never been asked. There haven't been very many performance related issues. I actually think gitlab might be the only service anyone has ever complained about and it actually used to be much slower! My recollection is that we addressed it with a combination of observation and common sense. For example, it stands to reason that decreasing the number of queries we make will improve performance.

I'm not sure there even is a truly objective way to measure performance. The bottleneck might be data transmission or server processing depending on your network and server speed.

Unless you have a better idea, my recommendation would be to run with `log_level = "DEBUG"` (or `log.level = DEBUG` if you're using ini configuration) in your [main section](https://bugwarrior.readthedocs.io/en/stable/common_configuration.html#main-section) and observe a `bugwarrior pull`. If that isn't sufficient maybe add some additional logging or timers.
